tests/5.1/target/test_target_memcpy_rect_async_depobj.c

```diff
@@ -0,0 +1,92 @@
+//===--- test_target_memcpy_rect_async_depobj.c ----------------------------===//
+//
+//  Inspired from OpenMP 5.1 Examples Doc, 5.16.4 & 8.9
+//  This test utilizes the omp_target_memcpy_rect_async construct to
+//  allocate memory on the device asynchronously. The construct
+//  uses 'obj' for dependency, so that memory is only copied once
+//  the variable listed in the depend clause is changed.
+//
+////===----------------------------------------------------------------------===//
+
+#include <omp.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "ompvv.h"
+#include <math.h>
+
+#define N 5
+
+int errors, i, j;
+
+int test_target_memcpy_async_depobj() {
+
+    int h, t;
+    errors = 0;
+    double *devRect;
+    h = omp_get_initial_device();
+    t = omp_get_default_device();
+
+    double* hostRect[N];
+    for(i = 0; i < N; i++){
+        hostRect[i] = (double *)malloc( sizeof(double)*N); 
+    } // 5x5 2D array
+    devRect = (double *)omp_target_alloc( sizeof(double)*N, t);
+    for(i = 0; i < N; i++){
+        devRect[i] = (double *)omp_target_alloc( sizeof(double)*N, t);
+    }
+
+    OMPVV_TEST_AND_SET_VERBOSE(errors, devRect == NULL);
+
+    for(i = 0; i < N; i++){
+        for (j = 0; j < N; j++){
+            hostRect[i][j] = i;
+        }
+    }
+    omp_depend_t obj;
+    #pragma omp depobj(obj) depend(inout: devRect)
+    omp_depend_t obj_arr[1] = {obj};
+
+    /* copy to device memory */
+    omp_target_memcpy_rect_async(devRect, hostRect, 
+                                32 /*number of bits??*/, 2, 5, //5 by 5
+                                0,          0,
+                                sizeof(double)*N*N, sizeof(double)*N*N,
+                                t,          h,
+                                1,          obj_arr);
```

Comment:
Fails here with:
```
test_target_memcpy_rect_async_depobj.c:51:61: warning: passing argument 5 of 'omp_target_memcpy_rect_async' makes pointer from integer without a cast [-Wint-conversion]
   51 |                                 32 /*number of bits??*/, 2, 5, //5 by 5
      |                                                             ^
```
The prototype is:
```C
int omp_target_memcpy_rect_async(
  void *dst,
  const void *src,
  size_t element_size,
  int num_dims,
  const size_t *volume,
  const size_t *dst_offsets,
  const size_t *src_offsets,
  const size_t *dst_dimensions,
  const size_t *src_dimensions,
  int dst_device_num,
  int src_device_num,
  int depobj_count,
  omp_depend_t *depobj_list
);
```
I think it should be something like the following:
```suggestion
    const size_t volume[2] = {5, 5};
    const size_t offsets[2] = {0, 0};
    const size_t dimensions[2] = {N, N};
    omp_target_memcpy_rect_async(devRect, hostRect,
                                 sizeof (double), 2,
                                 volume,
                                 offsets,    offsets,
                                 dimensions, dimensions,
                                 t,          h,
                                 1,          obj_arr);
```
And likewise below.

_Disclaimer: I have not made any attempt to check the bounds of the to-be transferred rectangle and whether it makes sense with the rest of the program._
_I have also not checked whether I interpreted the arguments correctly, but glancing at the spec, those should be okay._

Reply:
This seems correct, the use of arrays confused me at first i.e. "unsure of what the "dst/src_dimensions" should be, as well as what volume should be". I will use this w/ Dimensions as `{N, M}`